src/access/AccessRestriction.h

```diff
@@ -0,0 +1,288 @@
+    static constexpr size_t kRestrictionsPerEntry = CHIP_CONFIG_ACCESS_RESTRICTION_MAX_RESTRICTIONS_PER_ENTRY;
+
+    /**
+     * Defines the type of access restriction, which is used to determine the meaning of the restriction's id.
+     */
+    enum class Type : uint8_t
+    {
+        kAttributeAccessForbidden = 0,
+        kAttributeWriteForbidden  = 1,
+        kCommandForbidden         = 2,
+        kEventForbidden           = 3
+    };
+
+    /**
+     * Defines a single restriction on an attribute, command, or event.
+     *
+     * If id is not set, the restriction applies to all attributes, commands, or events of the given type (wildcard).
+     */
+    struct Restriction
+    {
+        Type restrictionType;
+        Optional<uint32_t> id;
+    };
+
+    /**
+     * Defines a single entry in the access restriction list, which contains a list of restrictions
+     * for a cluster on an endpoint.
+     */
+    struct Entry
+    {
+        FabricIndex fabricIndex;
+        EndpointId endpointNumber;
+        ClusterId clusterId;
+        std::vector<Restriction> restrictions;
+
+        bool IsValid() const
+        {
+            return endpointNumber != 0 && clusterId != app::Clusters::NetworkCommissioning::Id &&
+                clusterId != app::Clusters::Descriptor::Id;
+        }
+    };
+
+    /**
+     * Used to notify of changes in the access restriction list and active reviews.
+     */
+    class EntryListener
+    {
+    public:
+        enum class ChangeType
+        {
+            kAdded   = 1,
+            kRemoved = 2,
+            kUpdated = 3
+        };
+
+        virtual ~EntryListener() = default;
+
+        /**
+         * Notifies of a change in the access restriction list.
+         *
+         * @param [in] fabricIndex  The index of the fabric in which the entry has changed.
+         * @param [in] index        Index of entry to which has changed (relative to fabric).
+         * @param [in] entry        The latest value of the entry which changed.
+         * @param [in] changeType   The type of change that occurred.
+         */
+        virtual void OnEntryChanged(FabricIndex fabricIndex, size_t index, SharedPtr<Entry> entry, ChangeType changeType) = 0;
+
+        /**
+         * Notifies of an update to an active review with instructions and an optional redirect URL.
+         *
+         * @param [in] fabricIndex  The index of the fabric in which the entry has changed.
+         * @param [in] token        The token of the review being updated (obtained from ReviewFabricRestrictionsResponse)
+         * @param [in] instruction  The instructions to be displayed to the user.
+         * @param [in] redirectUrl  An optional URL to redirect the user to for more information.  May be null.
+         */
+        virtual void OnFabricRestrictionReviewUpdate(FabricIndex fabricIndex, uint64_t token, const char * instruction,
```

Comment:
`const char*` means that the backing be a null terminated string. Suggest CharSpan, which supports safe length handling and backing by different storage.

Reply:
done.